lib/openssl/asn1.rb

```diff
@@ -71,6 +106,91 @@ def initialize(value, tag, tag_class)
         @tag_class = tag_class
         @indefinite_length = false
       end
+
+      #
+      # call-seq:
+      #    asn1.to_der => DER-encoded String
+      #
+      # Encodes this ASN1Data into a DER-encoded String value. The result is
+      # DER-encoded except for the possibility of indefinite length forms.
+      # Indefinite length forms are not allowed in strict DER, so strictly speaking
+      # the result of such an encoding would be a BER-encoding.
+      #
+      def to_der
+        if @value.is_a?(Array)
+          cons_to_der
+        elsif @indefinite_length
+          raise ASN1Error, "indefinite length form cannot be used " \
+		        "with primitive encoding"
+        else
+          to_der_internal(der_value)
+        end
+      end
+
+      def der_value
+        raise TypeError, "no implicit conversion of #{self.class} into String" unless @value.respond_to?(:to_str)
+
+        String(@value)
+      end
+
+      private
+
+      def cons_to_der
+        ary = @value.to_a
+        str = "".b
+
+        @value.each_with_index do |item, idx|
+          if @indefinite_length && item.is_a?(EndOfContent)
+            if idx != ary.size - 1
+              raise ASN1Error, "illegal EOC octets in value"
+            end
+
+            break
+          end
+
+          item = item.to_der if item.respond_to?(:to_der)
+
+          str << item
+        end
+
+        to_der_internal(str, true)
+      end
+
+      def prim_to_der
+        to_der_internal(der_value)
+      end
+
+      def to_der_internal(body, constructed = false)
+        default_tag = ASN1.take_default_tag(self.class)
+        body_len = body.size
```

Comment:
`body`'s encoding isn't guaranteed to be ASCII-8BIT here, so we have to convert it.

This seems like another case where test coverage was not enough.

Reply:
if it's about this line, `.size` returns the correc bytesize, regardless of encoding.

Reply:
No, `String#size` is affected by the encoding. I think you meant `String#bytesize`.

Although for this line we could change it to `bytesize`, there is another issue with `String#<<` used later in this method, which require both String to be ASCII-compatible. For example, `"abc".b << "abc".encode("UTF-16LE")` doesn't work.

(FWIW there is an interesting feature request right now https://bugs.ruby-lang.org/issues/20594, it's unfortunate it's not there yet)

Reply:
I realized you added `String#b` call in `der_value`. It will also work.